src/backend/src/lib.rs

```diff
@@ -154,36 +155,63 @@ pub struct Config {
     pub ecdsa_key_name: String,
     // A list of allowed callers to restrict access to endpoints that do not particularly check or use the caller()
     pub allowed_callers: Vec<Principal>,
+    pub supported_credentials: Option<Vec<SupportedCredential>>,
+    /// Root of trust for checking canister signatures.
+    pub ic_root_key_raw: Option<Vec<u8>>,
+}
+
+fn update_config(arg: InitArg) {
+    mutate_state(|state| {
+        let ic_root_key_raw = match arg
+            .ic_root_key_der
+            .map(|der_key| extract_raw_root_pk_from_der(&der_key))
+        {
+            Some(Ok(root_key)) => Some(root_key),
+            Some(Err(_)) => None,
+            None => match extract_raw_root_pk_from_der(IC_ROOT_PK_DER) {
+                Ok(root_key) => Some(root_key),
+                Err(_) => None,
+            },
+        };
+        state
+            .config
+            .set(Some(Candid(Config {
+                ecdsa_key_name: arg.ecdsa_key_name,
+                allowed_callers: arg.allowed_callers,
+                supported_credentials: arg.supported_credentials,
+                ic_root_key_raw,
+            })))
+            .expect("setting config should succeed");
+    });
 }
 
 #[init]
 fn init(arg: Arg) {
     match arg {
-        Arg::Init(InitArg {
-            ecdsa_key_name,
-            allowed_callers,
-        }) => mutate_state(|state| {
-            state
-                .config
-                .set(Some(Candid(Config {
-                    ecdsa_key_name,
-                    allowed_callers,
-                })))
-                .expect("setting config should succeed");
-        }),
+        Arg::Init(arg) => update_config(arg),
         Arg::Upgrade => ic_cdk::trap("upgrade args in init"),
     }
 }
 
 #[post_upgrade]
-fn post_upgrade(_: Option<Arg>) {
-    read_state(|s| {
-        let _ = s
-            .config
-            .get()
-            .as_ref()
-            .expect("config is not initialized: reinstall the canister instead of upgrading");
-    });
+fn post_upgrade(arg: Option<Arg>) {
+    match arg {
+        Some(Arg::Init(arg)) => update_config(arg),
+        Some(Arg::Upgrade) => {
+            read_state(|s| {
+                let _ = s.config.get().as_ref().expect(
+                    "config is not initialized: reinstall the canister instead of upgrading",
+                );
+            });
+        }
+        None => {
```

Comment:
`None` is acceptable? If yes you can maybe spare some code duplication by replacing `Some(Arg::Upgrade)` and `None` with `_`.

If `None` is not acceptable, maybe trap?

Reply:
I believe it's ok yes. One could upgrade without an argument and we still want to check the config.

I'll use `_` thanks!